docs/tutorials/troubleshoot-a-course.md

```diff
@@ -0,0 +1,98 @@
+
+## Preface
+
+The easiest way to debug courses and assignments is to set up a foundation of continuous integration.
+If your course needs continuous integration, follow the [create continuous integration tutorial](create-continuous-integration.md).
+
+## Simple Fixes
+
+These fixes are quick sanity checks that catch simple issues.
+ - Ensure to install all necessary requirements to run the autograder.
+ - Upgrade your version of `autograder-py` by running:
+```
+pip3 install -U autograder-py
+```
+
+## Slow Graders
+
+As grading is the core user experience, fast graders are critical, but can be challenging.
+
+Every autograder-server instance has a maximum amount of time that a grading container can run before it is killed.
+The amount of time is determined by the [grading.runtime.max](../config.md#configuration-options) (system timeout) configuration on the autograder-server.
+
+Additionally, each assignment can set the [max-runtime-secs](../types.md#assignment) configuration option.
+The assignment max runtime cannot be greater than the system limit.
+
+Ensure these values are set as expected to avoid prematurely killing a grading container.
+Once you determine the number of seconds a grading container can run, check the following possibilities.
+
+### Infinite Loops / Recursion
+
+Submissions have the possibility to contain infinite loops or recursion.
+Graders should be prepared to detect when a question is taking too long to run, which can be quite difficult.
+
+We recommend setting a timeout for each question to kill the thread (or subprocess) that is grading the question.
+This allows graders to gracefully report slow (or infinite) implementations to the user.
+This way, the infinite loop only ruins a single question not the entire assignment.
+
+Be sure to test your graders properly timeout slow (or infinite) submissions.
+
+### Speeding Up Graders
+
+If ideal solutions are slow or cannot pass the tests within the grading container timeout,
+you will need to speed up grading scripts.
+
+While we cannot suggest language or grader based optimizations,
+we recommend the grading image complete as much work ahead of receiving a submission as possible.
+While this is not an exhaustive list of ideas, examples include:
+ - cloning a template repo
+ - caching assignment dependencies
+ - pre-compiling and linking code
+
+Leverage the `static-docker-commands` and `static-file-ops` to reduce post submission work for the grader.
+
+## Unexpected Grading Results
+
+When graders work locally but unexpectedly during docker grading,
+the easiest way to pinpoint issues is to manually step through the grading process inside the grading container.
+
+Using `assignment-01` from `my-first-course` from the [create-a-course tutorial](create-a-course.md) as an example,
+run the following (from the base directory (the one with the [README.md](../../README.md) in it)):
+```
+docker run -it --rm -v docs:/tmp/docs autograder.my-first-course.assignment-01 sh
```

Comment:
You should probably have them go through `docker images` to see what image to use.